### PR TITLE
Dftyaml

### PR DESCRIPTION
**Pull request by advikavs**
## Description

DRAFT PR on DFTYamlLoader

### CODE REVIEW

File: deepchem/data/data_loader.py [hunk 1]
```diff
@@ -1613,3 +1616,23 @@ def _featurize_shard(  # type: ignore[override]
             ids.append(entry_id)
         X = np.concatenate(features, axis=0)
         return X, np.array(labels), np.array(weights), np.array(ids)
+
+class DFTYamlLoader(DataLoader):
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1620]: Can you add the usage example and the docs?

File: deepchem/data/data_loader.py [hunk 2]
```diff
@@ -1613,3 +1616,23 @@ def _featurize_shard(  # type: ignore[override]
...
+    def __init__(self,
+                 featurizer: Featurizer):
+        tasks = None
+    def create_dataset(self, input_files: str, featurizer):
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1624]: Quick reminder on doc strings

File: deepchem/data/data_loader.py [hunk 3]
```diff
@@ -1613,3 +1616,23 @@ def _featurize_shard(  # type: ignore[override]
...
+        y = np.array([self._featurize_shard(shard) for shard in entries] )
+        return y 
+        
+    def _get_shards(self, input_files):
+        with open(input_files) as f:
+            data = yaml.load(f, Loader=SafeLoader)
+        return(data)
+    def _featurize_shard(self,
+                         shard):
+        e_type = shard['e_type']
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1635]: Make sure to document the required field in the docstring. Also, this should check if these fields are in the Yaml and throw an error if not so there is a good error message for a malformed input

File: deepchem/data/data_loader.py (hunk 1 above)
Comment (rbharath) [reply deepchem/data/data_loader.py:1620]: Add a full description of the expected file format for the YAML in this docstring. Remember that this is the main documentation for this filetype so it should have an example of the expected format and ideally references to other papers that use this format

File: docs/source/api_reference/data.rst [hunk 4]
```diff
@@ -151,6 +151,14 @@ On the other hand, the graph convolution models which inherit :code:`TorchModel`
 .. autoclass:: deepchem.feat.graph_data.GraphData
   :members:
```
Comment (rbharath) [inline docs/source/api_reference/data.rst:153]: Make sure to add the dft loader to the docs

File: deepchem/data/data_loader.py [hunk 5]
```diff
@@ -1613,3 +1616,23 @@ def _featurize_shard(  # type: ignore[override]
...
         X = np.concatenate(features, axis=0)
         return X, np.array(labels), np.array(weights), np.array(ids)
+
+class DFTYamlLoader(DataLoader):
+    def __init__(self,
+                 featurizer: Featurizer):
+        tasks = None
+    def create_dataset(self, input_files: str, featurizer):
+        entries = self._get_shards(input_files)
+        y = np.array([self._featurize_shard(shard) for shard in entries] )
```
Comment (rbharath) [inline deepchem/data/data_loader.py:1626]: This is not correct; a `DataLoader` should return a DeepChem `Dataset` object and not a numpy array. Check the other data loaders to check their return types

File: deepchem/data/tests/test_dftyaml.py [hunk 6]
```diff
@@ -0,0 +1,9 @@
+from deepchem.data.data_loader import DFTYamlLoader
+import deepchem as dc
+
+def test_dftloader():
+    input_files = 'dftdata.yaml'
+    featurizer = dc.feat.DummyFeaturizer()
+    k = DFTYamlLoader(featurizer)
+    l = k.create_dataset(input_files, featurizer)
```
Comment (rbharath) [inline deepchem/data/tests/test_dftyaml.py:8]: There should be a check here this returns a DeepChem `Dataset` object

File: deepchem/data/tests/test_dftyaml.py [hunk 7]
```diff
@@ -0,0 +1,9 @@
+from deepchem.data.data_loader import DFTYamlLoader
+import deepchem as dc
+
+def test_dftloader():
```
Comment (rbharath) [inline deepchem/data/tests/test_dftyaml.py:4]: The test file has 2 entries. There should be a check both entries are loaded correctly

**Review by rbharath — COMMENTED** [review]
Good start! Biggest point to address is that the data loader should return a DeepChem dataset and not a numpy array. There will also need to be a lot of documentation added to fully specify the file format for these systems.

File: deepchem/data/data_loader.py (hunk 5 above)
Comment (advikavs) [inline deepchem/data/data_loader.py:1626]: name it X instead of y

**Pull request closed without merging**